### DISCUSSION TITLE
Difference between virtink and kata？

### DISCUSSION THREAD

**Discussion by daxiguafan** [smartxworks/virtink#52] [Q&A] [answered]
As i konw, both use container img as vm's rootfs. Is there advantage for using virtink?

**Comment by fengye87** [comment, marked as answer]
As far as I know about Kata containers, it integrates with Kubernetes in the CRI level, which means all Pods would run in a virtualized box. As for Virtink, it integrates with Kubernetes in the CRD and controller level. Pods are still Pods, run by Docker or containerd or other non-virtualization CRIs, while VMs are introduced to Kubernetes in an addon fashion. For each VM, Virtink would first create a standard Pod. Then inside the Pod, Virtink runs the VM process, and make the Pod's volumes and network available to the VM. So to sum up, Kata containers tries to make Pods run more safely and isolated, while Virtink tries to make VM a new citizen in Kubernetes along side with canonical Pods. @daxiguafan

**Reply by ccrvlh** [reply]
That's really interesting, and thanks for sharing this @fengye87. With new MicroVM hypervisors (Firecracker/CH), having VM's as first class citizen is definitely a great opportunity for K8S, and virtink does fill an interesting gap for low overhead virtualization that Kubevirt doesn't, while Kata brings the isolation with limited flexibility (kernel/ base vm image for example) at the VM level.

Wondering if you guys considered looking into some integration at the kernel/rootfs level like unikernels/linuxkit? After playing around with a few projects, my impression is that what is lacking the most (apart from having lightweight VMs at the control plane level that this project solves) is lightweight kernel + applications. A simple Redis workload for example that has  almost zero overhead for the containers, and ~2-3 MB while running (no data), uses 120MB extra when adding the Kernel + OS with Kata (with Cloud Hypervisor), extremely expensive. gVisor does a great job giving a great isolation by protecting syscalls with a very low overhead (~10MB) from my tests with Nginx/Redis images. But does have challenges, specially with performance.

Apart from that, would you have some more documentation on the architecture and design decisions, maybe a contributing guide? Thanks a lot for sharing this, the project looks great, looking forward to spend more time testing and tinkering with it.

**Reply by fengye87** [reply]
Thank you @lowercase00 . We haven't taken a serious look into unikernels/linuxkit, but is definitely an interesting topic to explore. It probably deserves its own project since the use cases and design goals could be quite different.

And yes, the project is currently lacking a lot of docs. Hopefully we would bring more docs in the first quarter of the new year, right after we finish features like disk hot plugging and IPv6 support.
